src/cfi.rs

```diff
@@ -0,0 +1,938 @@
+        assert_eq!(CommonInformationEntry::parse(EndianBuf::<LittleEndian>::new(&contents)),
+                   Err(Error::UnexpectedEof));
+    }
+
+    #[test]
+    fn test_parse_cie_incomplete_length_64() {
+        let section = Section::with_endian(Endian::Little)
+            .L32(0xffffffff)
+            .L32(12345);
+        let contents = section.get_contents().unwrap();
+        assert_eq!(CommonInformationEntry::parse(EndianBuf::<LittleEndian>::new(&contents)),
+                   Err(Error::UnexpectedEof));
+    }
+
+    #[test]
+    fn test_parse_cie_incomplete_id_32() {
+        let section = Section::with_endian(Endian::Big)
+            // The length is not large enough to contain the ID.
+            .B32(3)
+            .B32(0xffffffff);
+        let contents = section.get_contents().unwrap();
+        assert_eq!(CommonInformationEntry::parse(EndianBuf::<BigEndian>::new(&contents)),
+                   Err(Error::UnexpectedEof));
+    }
+
+    #[test]
+    fn test_parse_cie_bad_id_32() {
+        let section = Section::with_endian(Endian::Big)
+            // Initial length
+            .B32(4)
+            // Not the CIE Id.
+            .B32(0xbad1bad2);
+        let contents = section.get_contents().unwrap();
+        assert_eq!(CommonInformationEntry::parse(EndianBuf::<BigEndian>::new(&contents)),
+                   Err(Error::NotCieId));
+    }
+
+    #[test]
+    fn test_parse_cie_32_bad_version() {
+        let length = Label::new();
+        let start = Label::new();
+        let end = Label::new();
+
+        let section = Section::with_endian(Endian::Little)
+            // Initial length
+            .L32(&length)
+            .mark(&start)
+            // CIE Id
+            .L32(0xffffffff)
+            .cie_header(99,
+                        None,
+                        4,
+                        0,
+                        1,
+                        2,
+                        3)
+            .mark(&end);
+
+        let expected_length = (&end - &start) as u64;
+        length.set_const(expected_length);
+
+        let contents = section.get_contents().unwrap();
+
+        assert_eq!(CommonInformationEntry::parse(EndianBuf::<LittleEndian>::new(&contents)),
+                   Err(Error::UnknownVersion));
+    }
+
+    #[test]
+    fn test_parse_cie_unknown_augmentation() {
+        let length = Label::new();
+        let start = Label::new();
+        let end = Label::new();
+
+        let augmentation = Some("replicant");
+        let expected_rest = [1, 2, 3];
+
+        let section = Section::with_endian(Endian::Little)
+            // Initial length
+            .L32(&length)
+            .mark(&start)
+            // CIE Id
+            .L32(0xffffffff)
+            .cie_header(4,
+                        augmentation,
+                        4,
+                        0,
+                        1,
+                        2,
+                        3)
```

Comment:
This is an improvement over byte arrays, but it is still too verbose and repetitive.  How about changing `cie_header` to take the expected `CommonInformationEntry` instead (we'll still have to fill in the length afterwards).